File: runtime/typescript/packages/core/src/model/discovery.ts
import * as fs from "fs";

import { ModelInfo } from "./contracts/models/model-info.js";

type CapabilityEntry = {
  prefix?: unknown;
  contextWindow?: unknown;
  inputModalities?: unknown;
  outputModalities?: unknown;
};

type CapabilityDataset = {
  providers?: Record<string, CapabilityEntry[]>;
};

let cachedDataset: CapabilityDataset | undefined;

function getDataset(): CapabilityDataset {
  if (cachedDataset === undefined) {
    const contents = fs.readFileSync(new URL("./model_capabilities.json", import.meta.url), "utf-8");
    cachedDataset = JSON.parse(contents) as CapabilityDataset;
  }
  return cachedDataset;
}

function isTokenBoundary(modelId: string, prefix: string): boolean {
  return modelId === prefix || (modelId.startsWith(prefix) && !/[A-Za-z0-9]/.test(modelId[prefix.length] ?? ""));
}

function asRecord(value: unknown): Record<string, unknown> {
  if (value !== null && typeof value === "object" && !Array.isArray(value)) {
    return { ...(value as Record<string, unknown>) };
  }
  return {};
}

function setStringField(target: ModelInfo, field: "id" | "displayName" | "ownedBy", value: unknown): void {
  if (value !== undefined && value !== null) {
    target[field] = String(value);
  }
}

function setNumberField(target: ModelInfo, field: "contextWindow", value: unknown): void {
  if (value !== undefined && value !== null) {
    target[field] = Number(value);
  }
}

function setStringArrayField(
  target: ModelInfo,
  field: "inputModalities" | "outputModalities",
  value: unknown,
): void {
  if (value !== undefined && value !== null) {
    target[field] = Array.isArray(value) ? value.map(item => String(item)) : [];
  }
}

export function matchCapabilities(modelId: string, provider: string): CapabilityEntry | null {
  const entries = getDataset().providers?.[provider];
  if (!entries) {
    return null;
  }

  let best: CapabilityEntry | null = null;
  let bestLength = -1;
  for (const entry of entries) {
    if (typeof entry.prefix !== "string") {
      continue;
    }
    if (entry.prefix.length > bestLength && isTokenBoundary(modelId, entry.prefix)) {
      best = entry;
      bestLength = entry.prefix.length;
    }
  }
  return best;
}

export function enrich(base: ModelInfo, provider: string): ModelInfo {
  const entry = matchCapabilities(base.id, provider);
  if (entry === null) {
    return base;
  }

  if (base.contextWindow === undefined || base.contextWindow === null) {
    if (entry.contextWindow !== undefined && entry.contextWindow !== null) {
      base.contextWindow = structuredClone(entry.contextWindow) as number;
    }
  }
  if (base.inputModalities === undefined || base.inputModalities === null) {
    if (entry.inputModalities !== undefined && entry.inputModalities !== null) {
      base.inputModalities = structuredClone(entry.inputModalities) as string[];
    }
  }
  if (base.outputModalities === undefined || base.outputModalities === null) {
    if (entry.outputModalities !== undefined && entry.outputModalities !== null) {
      base.outputModalities = structuredClone(entry.outputModalities) as string[];
    }
  }
  return base;
}

export function mapModel(raw: unknown, provider: string): ModelInfo {
  const data = asRecord(raw);
  const info = new ModelInfo();
  info.additionalProperties = data;

  switch (provider) {
    case "anthropic":
      setStringField(info, "id", data.id);
      setStringField(info, "displayName", data.display_name);
      info.ownedBy = "anthropic";
      setNumberField(info, "contextWindow", data.context_length);
      setStringArrayField(info, "inputModalities", data.input_modalities);
      setStringArrayField(info, "outputModalities", data.output_modalities);
      break;
    case "foundry":
      if ("properties" in data) {
        const props = asRecord(data.properties);
        const model = asRecord(props.model);
        const caps = asRecord(props.capabilities);
        setStringField(info, "id", data.name);
        setStringField(info, "displayName", model.name);
        setStringField(info, "ownedBy", model.publisher);
        setNumberField(info, "contextWindow", model.maxContextLength);
        setStringArrayField(info, "inputModalities", caps.supportedInputModalities);
        setStringArrayField(info, "outputModalities", caps.supportedOutputModalities);
      } else if ("modelName" in data || data.type === "ModelDeployment") {
        setStringField(info, "id", data.name);
        setStringField(info, "displayName", data.modelName);
        setStringField(info, "ownedBy", data.modelPublisher);
        setNumberField(info, "contextWindow", data.maxContextLength);
      } else {
        setStringField(info, "id", data.id);
        setStringField(info, "ownedBy", data.owned_by);
        setNumberField(info, "contextWindow", data.maxContextLength);
      }
      break;
    default:
      setStringField(info, "id", data.id);
      setStringField(info, "ownedBy", data.owned_by);
      break;
  }

  if (info.id === undefined || info.id === null) {
    info.id = "";
  }
  return info;
}

File: runtime/python/prompty/prompty/providers/discovery.py
"""Shared, cross-provider model-discovery algorithms.

This module owns the two behaviors validated by the ``DiscoveryConformance``
vectors — identically across every Prompty runtime:

* :func:`map_model` — map a raw provider model / deployment / catalog payload
  into a canonical :class:`~prompty.model.ModelInfo`.
* :func:`enrich` — fill *only* the missing capability fields of a sparse
  ``ModelInfo`` from the shared, embedded capability dataset
  (``model_capabilities.json``), using longest-prefix, token-boundary matching.

The dataset is deliberately NOT emitted from TypeSpec: it is volatile provider
data (context windows, modalities, new model families) refreshed as a snapshot.
Every runtime embeds a copy of the same file and applies the same rule:
provider-supplied fields always win; dataset entries fill only-missing fields.
"""

from __future__ import annotations

import json
from copy import deepcopy
from functools import lru_cache
from pathlib import Path
from typing import Any

from ..model import ModelInfo

__all__ = ["map_model", "enrich", "match_capabilities"]

_DATASET_PATH = Path(__file__).parent / "model_capabilities.json"

# Capability fields that :func:`enrich` fills from the dataset. These are the
# only fields the dataset carries; ``id`` / ``displayName`` / ``ownedBy`` are
# always provider-supplied.
_CAPABILITY_FIELDS = ("context_window", "input_modalities", "output_modalities")

# Mapping from ModelInfo attribute name -> dataset (camelCase) key.
_DATASET_KEYS = {
    "context_window": "contextWindow",
    "input_modalities": "inputModalities",
    "output_modalities": "outputModalities",
}


@lru_cache(maxsize=1)
def _dataset() -> dict[str, Any]:
    """Load and cache the embedded capability dataset."""
    with _DATASET_PATH.open(encoding="utf-8") as handle:
        return json.load(handle)


def _is_token_boundary(model_id: str, prefix: str) -> bool:
    """Return True if ``prefix`` ends at a token boundary within ``model_id``.

    A prefix matches only when it is either the whole id or is immediately
    followed by a non-alphanumeric separator (``-``, ``:``, ``.``, ``/`` ...).
    This prevents ``gpt-4`` from matching ``gpt-45-future`` while still
    matching ``gpt-4o-mini-2024-07-18`` against ``gpt-4o-mini``.
    """
    if model_id == prefix:
        return True
    if not model_id.startswith(prefix):
        return False
    return not model_id[len(prefix)].isalnum()


def match_capabilities(model_id: str, provider: str) -> dict[str, Any] | None:
    """Return the longest-prefix dataset entry for ``model_id`` under ``provider``.

    Returns ``None`` when the provider has no dataset entries or no prefix
    matches at a token boundary.
    """
    entries = _dataset().get("providers", {}).get(provider)
    if not entries:
        return None
    best: dict[str, Any] | None = None
    best_len = -1
    for entry in entries:
        prefix = entry.get("prefix", "")
        if _is_token_boundary(model_id, prefix) and len(prefix) > best_len:
            best = entry
            best_len = len(prefix)
    return best


def enrich(base: ModelInfo, provider: str) -> ModelInfo:
    """Fill only-missing ``ModelInfo`` capability fields from the dataset.

    Provider-supplied values always win: a field is filled only when it is
    ``None`` on ``base``. A present-but-empty list (e.g. ``[]`` for an
    embedding model's output modalities) is an intentional provider signal and
    is preserved. Mutates and returns ``base``.
    """
    entry = match_capabilities(base.id, provider)
    if entry is None:
        return base
    for attr in _CAPABILITY_FIELDS:
        if getattr(base, attr) is None:
            key = _DATASET_KEYS[attr]
            if key in entry:
                # Copy so a downstream mutation of the ModelInfo cannot corrupt
                # the lru_cached dataset (lists are shared references otherwise).
                setattr(base, attr, deepcopy(entry[key]))
    return base


def _as_dict(raw: Any) -> dict[str, Any]:
    """Coerce a raw provider payload (dict or SDK object) into a plain dict."""
    if isinstance(raw, dict):
        return raw
    for method in ("model_dump", "to_dict", "dict"):
        fn = getattr(raw, method, None)
        if callable(fn):
            try:
                result = fn()
            except TypeError:
                continue
            if isinstance(result, dict):
                return result
    data = getattr(raw, "__dict__", None)
    if isinstance(data, dict):
        return dict(data)
    return {}


def map_model(raw: Any, provider: str) -> ModelInfo:
    """Map a raw provider payload to a canonical :class:`ModelInfo`.

    Dispatches on ``provider`` to handle each provider's payload shape. The
    entire raw payload is echoed verbatim into ``additional_properties`` so no
    provider-specific field is lost. Absent fields remain ``None`` (and are
    therefore omitted by ``ModelInfo.save()``).
    """
    data = _as_dict(raw)
    info = ModelInfo()
    info.additional_properties = data

    if provider == "anthropic":
        info.id = data.get("id")
        info.display_name = data.get("display_name")
        info.owned_by = "anthropic"
        info.context_window = data.get("context_length")
        info.input_modalities = data.get("input_modalities")
        info.output_modalities = data.get("output_modalities")
    elif provider == "foundry":
        if "properties" in data:  # nested ARM deployment
            props = data.get("properties") or {}
            model = props.get("model") or {}
            caps = props.get("capabilities") or {}
            info.id = data.get("name")
            info.display_name = model.get("name")
            info.owned_by = model.get("publisher")
            info.context_window = model.get("maxContextLength")
            info.input_modalities = caps.get("supportedInputModalities")
            info.output_modalities = caps.get("supportedOutputModalities")
        elif "modelName" in data or data.get("type") == "ModelDeployment":  # flat deployment
            info.id = data.get("name")
            info.display_name = data.get("modelName")
            info.owned_by = data.get("modelPublisher")
            info.context_window = data.get("maxContextLength")
        else:  # catalog model
            info.id = data.get("id")
            info.owned_by = data.get("owned_by")
            info.context_window = data.get("maxContextLength")
    else:  # openai (default)
        info.id = data.get("id")
        info.owned_by = data.get("owned_by")

    if info.id is None:
        info.id = ""
    return info
